### PR TITLE
Allow to run unit tests using KUnit with a user-space like syntax

### PR DESCRIPTION
**Pull request by JoseExposito**
## Why?

I'm working on the Rust abstractions for the HID subsystem and on a HID driver.

In order to be able to test the HID subsystem abstractions, I need to avoid calling the C APIs, so I created a bindings mock module that I import as:

```rust
#[cfg(CONFIG_KUNIT)]
use crate::hid::bindings_mock as bindings;
```

Which returns well known values for my tests. For reference, this is how the driver adapter test looks like:

```rust
#[kunit_tests(rust_kernel_hid_driver)]
mod tests {
    use super::*;
    use crate::{c_str, driver, hid, prelude::*};
    use core::ptr;

    struct SimpleTestDriver;
    impl Driver for SimpleTestDriver {
        type Data = ();
    }

    #[test]
    fn rust_test_hid_driver_adapter() {
        let mut hid = bindings::hid_driver::default();
        let name = c_str!("SimpleTestDriver");
        static MODULE: ThisModule = unsafe { ThisModule::from_ptr(ptr::null_mut()) };

        let res = unsafe {
            <hid::Adapter<SimpleTestDriver> as driver::DriverOps>::register(&mut hid, name, &MODULE)
        };
        assert_eq!(res, Err(ENODEV)); // The mock returns -19
    }
}
```

While I could do the same using documentation tests, using regular tests allows to reduce code duplication.

## About the patches

The first patch introduces the low level macros `kunit_case!` and `kunit_test_suite!`. It is based on the work mentioned by @sulix in https://github.com/Rust-for-Linux/linux/pull/935#issuecomment-1338939583. David, I added you as Co-developed-by, let me know if that works for you, please.

The second patch introduces a procedural macro that allows to run the tests almost with the same syntax used in user-space. The only change required is replacing `#[cfg(test)]` with `#[kunit_tests(kunit_test_suit_name)]`.

## Possible issues/questions

- ~~The series relies on `#[cfg(CONFIG_KUNIT)]` being defined only while running the tests. A similar approach is used by [static stub patches](https://lore.kernel.org/linux-kselftest/20221208061841.2186447-2-davidgow@google.com/T/), so I hope that it also works here, but it'd be nice to hear David's opinion.~~
 Thanks to David for clarifying this point, it is solved now.

- As far as I know, procedural macros don't support `$crate` and I was not able to find a workaround to be able to use this macro both by the kernel crate and in drivers. In the macro, I use code like `crate::bindings::<...>` that won't work when used by a driver. Do you know if it is possible to fix this?

### CODE REVIEW

**Comment by sulix** [comment]
Thanks very much for doing this — it worked fine for me here. 

A few comments:

> The first patch introduces the low level macros `kunit_case!` and `kunit_test_suite!`. It is based on the work mentioned by @sulix in [#935 (comment)](https://github.com/Rust-for-Linux/linux/pull/935#issuecomment-1338939583). David, I added you as Co-developed-by, let me know if that works for you, please.

Works fine for me: thanks for tidying it all up!

> The second patch introduces a procedural macro that allows to run the tests almost with the same syntax used in user-space. The only change required is replacing `#[cfg(test)]` with `#[kunit_tests(kunit_test_suit_name)]`.

While I'm not familiar enough to comment on proc macro implementation, I quite like the syntax here personally.

> The series relies on `#[cfg(CONFIG_KUNIT)]` being defined only while running the tests. A similar approach is used by [static stub patches](https://lore.kernel.org/linux-kselftest/20221208061841.2186447-2-davidgow@google.com/T/), so I hope that it also works here, but it'd be nice to hear David's opinion.

I'm afraid that's not something you can rely on in all circumstances. There are some setups which unconditionally compile KUnit as a module (I think Red Hat and Android are going down this path), as well as the possibility of wanting to run the tests and the real driver on the same system.

The "proper" way to check if a test is running is by using the combination of CONFIG_KUNIT and checking current->kunit_test is not NULL. This allows, e.g., stubs to only be active on KUnit's threads.

The current version of the stubs implementation (which you linked) does not work if CONFIG_KUNIT=m, but that should be fixed in the next revision.

tl;dr: If CONFIG_KUNIT is not set, there are definitely no tests, but it's not true to say that CONFIG_KUNIT is _only_ set while tests are running: it's perfectly possible (even if it's, IMHO, a terrible idea) to set CONFIG_KUNIT=y and then use a kernel normally.

Further thoughts (which could be relegated to follow-up patches if you prefer) include:
- Could we include a sample test to make sure these work? Something simple this would do to start (and more tests could be added as more features are supported):
```
#[kunit_tests(kunit_rust)]
mod tests {
    #[test]
    fn example_test() {
        assert_eq!(1+1, 2);
    }
}
```
- It'd be nice to expose the KUnit assert/expect macros as an alternative to Rust's built-in ones. The KUnit tooling doesn't handle the failed rust assertions properly, treating it as a crash. And expectations are probably nicer, as exiting a test early is not handled well with KUnit/Rust (see #759).
- Unlike with the doctests' implementation, I don't see a way to access the ``struct kunit`` object. I suspect that using ``current->kunit_test`` is the right way to handle that.
- It may make sense to start to use some of these macros for (and otherwise share code with) the doctests implementation, too.

Regardless, this is looking pretty good to me, thanks!

File: rust/kernel/kunit.rs [hunk 1]
```diff
@@ -88,3 +88,89 @@ macro_rules! kunit_assert_eq {
...
+            let mut ret = [0; 256];
+
+            let mut i = 0;
+            while i < name_u8.len() {
+                ret[i] = name_u8[i] as i8;
+                i += 1;
+            }
+
+            ret
+        };
```
Comment (sulix) [inline rust/kernel/kunit.rs]: It may make more sense for us to just outline the suite name generally, so that c_str!() and similar can be used (and we could get rid of the length limit as well). That'd require changes to the C code as well, but I don't think that'd be a problem.

File: rust/macros/kunit.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    // static mut TEST_CASES : &mut[crate::bindings::kunit_case] = unsafe {
+    //     &mut [KUNIT_CASE_FOO, KUNIT_CASE_BAR, KUNIT_CASE_NULL]
+    // };
+    //
+    // crate::kunit_test_suite!(kunit_test_suit_name, TEST_CASES);
+    // ```
+    let mut kunit_macros = "".to_owned();
+    let mut test_cases = "".to_owned();
+    for test in tests {
+        let kunit_stub_fn_name = format!("kunit_stub_{}", test);
```
Comment (sulix) [inline rust/macros/kunit.rs]: To avoid confusion with the static stub feature, could we maybe give this a different name?

Maybe "thunk" instead?

Equally, would it make sense to include "rust" somewhere in the name? I could go either way on that, but it might make it clearer…

**Comment by JoseExposito** [comment]
Hi David,

Thanks a lot for your comments! You really helped me understand some points.

> tl;dr: If CONFIG_KUNIT is not set, there are definitely no tests, but it's not true to say that CONFIG_KUNIT is only set while tests are running: it's perfectly possible (even if it's, IMHO, a terrible idea) to set CONFIG_KUNIT=y and then use a kernel normally.

While this doesn't affect this merge request, it complicates a bit creating mocks.

It means that we won't be able to remove the mock code at compile time, introducing a bit of run-time overhead checking for `current->kunit_test` :(

Next weekend, I'll try to add an example of how mocking a module could work in Rust. Hopefully it'll make easier to understand the problematic.

> Could we include a sample test to make sure these work? 

Sure, good point.

> It'd be nice to expose the KUnit assert/expect macros as an alternative to Rust's built-in ones. The KUnit tooling doesn't handle the failed rust assertions properly, treating it as a crash. And expectations are probably nicer, as exiting a test early is not handled well with KUnit/Rust (see How to handle Kunit killing the current thread on assertion failures? #759).

I could try to have a look into that in the future,. Hopefully, it shouldn't be too complicated.
But I'd like to finish the driver that originated this MR first. Given that I do kernel stuff mostly on the weekends, it could take me a while...

> Unlike with the doctests' implementation, I don't see a way to access the struct kunit object. I suspect that using current->kunit_test is the right way to handle that.

I removed the test parameter from the signature on purpose to keep test cases as similar as possible to user-space.

I think that using `current->kunit_test` is the best option as well. Ideally, the KUnit assert/expect Rust macros should get `current->kunit_test` hiding the extra complexity.


Thanks again for reviewing this! I'll address your comments as soons as possible as submit v2.

File: rust/kernel/kunit.rs (hunk 1 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs:157]: Good point, I renamed it to `kunit_rust_wrapper_` to make sure it is 100% clear what this function is.

File: rust/macros/kunit.rs (hunk 2 above)
Comment (JoseExposito) [reply rust/macros/kunit.rs]: Good point, I renamed it to `kunit_rust_wrapper_` to make sure it is 100% clear what this function is.

File: rust/kernel/kunit.rs (hunk 1 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs]: Yes, changing the name field from `char[256]` to `char *` would simplify this code a little bit. In fact (probably because I'm a Rust noob) casting this field was the most time consuming part of this PR :sweat_smile: 

For the sake of simplicity, I don't mind keeping this code. Changing C code (upstream and here) would require backporting and might slow down the inclusion of this PR. But if you want to change the field type, let me know and I'll submit a patch upstream.

**Comment by JoseExposito** [comment]
Hi @sulix,

I just pushed v2 of the patches.

In addition to addressing your comments, this second version introduces a new patch adding the `is_kunit_test()` function.

This function checks `CONFIG_KUNIT` and `current->kunit_test` to decide if we are in a KUnit test execution or not. I noticed that in your stub series you are using `kunit_get_current_test()`, which is faster.

Unfortunately, `kunit_get_current_test()` is not available yet in the Rust branch, so for the moment I'll kept it as it is and once the changes from upstream are backported I'll update my function.

Also, I included a couple of examples/Kunit tests to show how to mock a function and an entire module. The module mocked in the example is `bindings`, a module most abstractions developers would be interested in mocking.

Thanks again for looking into this,
Jose

File: rust/kernel/kunit.rs [hunk 3]
```diff
@@ -88,3 +92,173 @@ macro_rules! kunit_assert_eq {
...
+/// // This is the function we want to test. Since `bindings` has been mocked, we can use its
+/// // functions seamlessly.
+/// fn get_boot_ns() -> u64 {
+///     unsafe { bindings::ktime_get_boot_fast_ns() }
+/// }
+///
+/// let time = get_boot_ns();
+/// assert_eq!(time, 1234);
+/// ```
+pub fn is_kunit_test() -> bool {
```
Comment (sulix) [inline rust/kernel/kunit.rs]: I'd personally prefer the name ``in_kunit_test()``, as it may make sense to call this (or future related functions) from code which is not _part of_ the test, but is nevertheless run in a test context (e.g., for KASAN, we check if a test is running when making a report).

File: rust/kernel/kunit.rs (hunk 1 above)
Comment (sulix) [reply rust/kernel/kunit.rs]: Up to you: I'm happy to accept a change upstream to change this if you'd prefer. Otherwise, it's something we can either leave as-is or change later, once everything's accepted upstream.

I'd (optimistically) assumed that very little rust stuff would be backported to earlier kernel versions, so as long as we can stay relatively in-sync between the Rust tree and the KUnit one, it shouldn't be _too_ horrific. If you're planning to either backport to much older releases, or to try to push a driver depending on this via another branch in the short term, though, that could be trickier…

File: rust/kernel/kunit.rs [hunk 4]
```diff
@@ -88,3 +92,173 @@ macro_rules! kunit_assert_eq {
...
+
+        // SAFETY: `KUNIT_TEST_SUITE` is static.
+        #[used]
+        #[link_section = ".kunit_test_suites"]
+        static mut KUNIT_TEST_SUITE_ENTRY: *const $crate::bindings::kunit_suite =
+            unsafe { &KUNIT_TEST_SUITE };
+    };
+}
+
+/// In some cases, you need to call test-only code from outside the test case, for example, to
```
Comment (sulix) [inline rust/kernel/kunit.rs:188]: We may want to distinguish between this sort of "mocking", which only replaces direct calls with other KUnit mocking functionality, like the upcoming ``static stub`` feature:
https://lore.kernel.org/linux-kselftest/20221208061841.2186447-1-davidgow@google.com/

The main differences are that the static stub feature redirects _all_ calls to a function, even if it's indirect (e.g., from a different module, crate, file, etc). It also automatically does the in_kunit_test() check, and the replacement can be swapped in and out at runtime.

To be clear, I think this is fine as-is, but we will want to come up with a consistent story about when to use which features (and probably a more consistent naming scheme) once the static stub patches land.

In the meantime, we've used the term "function redirection" to refer to the general concept of replacing a function, and then had more specific names for different implementations (mostly "stubbing"-related). A [somewhat-outdated document](https://kunit.dev/mocking.html) gives some of the rationale behind this, mostly that "mocking" as a term is a bit overloaded.

File: rust/kernel/kunit.rs [hunk 5]
```diff
@@ -174,3 +176,12 @@ macro_rules! kunit_test_suite {
             unsafe { &KUNIT_TEST_SUITE };
     };
 }
+
+#[kunit_tests(rust_kernel_kunit)]
+mod tests {
+    #[test]
+    fn rust_test_kunit_kunit_tests() {
+        let running = true;
+        assert_eq!(running, true);
```
Comment (sulix) [inline rust/kernel/kunit.rs:263]: Again, I think that someday we'll want to be able to use the KUNIT assertion system (and perhaps encourage more use of expectations over assertions, due to the [issues with stack unwinding](https://github.com/Rust-for-Linux/linux/issues/759)), but I think this can definitely be a separate patch.

**Review by sulix — APPROVED** [review]
Thanks!

I'm pretty happy with this. The only real change I'd suggest is renaming is_kunit_test() to in_kunit_test(), to make it clear that we're running in a test context. (But if there's a strong reason to keep is_kunit_test(), I could live with it…)

Otherwise, this looks pretty good. I'm not going to pretend to be enough of a rust expert to have caught issues with all of the code properly (particularly the proc macros), but from a KUnit point of view, I'm happy with it.

(One, slightly related note for the future, is that we're going to add a more generic way of having "hooks" where non-test code calls into KUnit, à la kunit_get_current_test()/in_kunit_test(). This is [pretty macro heavy](https://lore.kernel.org/linux-kselftest/20230124080350.2275652-1-davidgow@google.com/) at the moment, so if it's going to make doing future things in rust really difficult (due to, e.g., bindgen), feel free to complain / suggest changes. It's more a problem for a future patch, though (and I suspect it'd need wrapping / redoing to be more idiomatic anyway), so it's not really a problem for now.)

File: rust/kernel/kunit.rs (hunk 1 above)
Comment (ojeda) [reply rust/kernel/kunit.rs]: I am confused -- originally I thought José meant back merges, i.e. to bring the C change into the `rust` branch here. That is something I did before the merge every few weeks, and we were discussing the other day whether to continue doing it until the tree is ready to be decommissioned.

But then David mentions backporting to older release versions: indeed, that is not something we plan to do. (Not that I wouldn't be glad to be in a situation where we could think about it -- it would mean Rust was wildly successful in the kernel :)

Which one of the two are we discussing?

**Comment by ojeda** [comment]
> This is [pretty macro heavy](https://lore.kernel.org/linux-kselftest/20230124080350.2275652-1-davidgow@google.com/) at the moment, so if it's going to make doing future things in rust really difficult (due to, e.g., bindgen), feel free to complain / suggest changes.

I don't think we need to constraint the C side here. So unless it is a blocker, or something that is anyway an improvement for both sides (like the `const` correctness one I sent once), I would say we should try to find the best way for the C side, and then see if we can work with that.

File: rust/kernel/kunit.rs (hunk 3 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs]: Renamed, I like `in_` more as well

File: rust/kernel/kunit.rs (hunk 4 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs:188]: I added an entry in my to-do list to document mocking and add a link to the general KUnit docs in https://github.com/Rust-for-Linux/linux/pull/935 once this PR gets merged.

At work, I write JavaScript and the terminology makes sense to me. It's similar to the one used by the popular testing frameworks.

File: rust/kernel/kunit.rs (hunk 1 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs]: > I am confused -- originally I thought José meant back merges, i.e. to bring the C change into the rust branch here.

This is what I meant. While applying the patch both upstream and here is trivially easy, I'd prefer to keep this PR as simple as possible and avoid unnecessary interdependence.

I'm working a more complex series based on this PR and, hopefully, keeping things as simple as possible, will help getting it merged.

Comment (sulix) [reply rust/kernel/kunit.rs]: Yeah: I was a bit unsure which was meant, so tried to address both.

How about we take this PR in as-is (with the char[256] array), and when it lands in Linus' tree, we can decide if changing it is worthwhile? That'll avoid any complications.

File: rust/kernel/kunit.rs (hunk 4 above)
Comment (sulix) [reply rust/kernel/kunit.rs:188]: Yeah: personally, I'd rather we focus this PR on getting the basics of running tests going, and leave any extensive documentation of mocking or related things to a separate one.

Terminology-wise, I'm not too worried about matching that old document, which was mostly based on an abandoned "mocking" proposal which had some more advanced "mocking" features, like autogenerating stubs, and being able to assert easily that a given call will receive certain arguments. Since that was abandoned, and the "stubs" system (which is much simpler) hasn't quite landed yet, none of the nomenclature here is set in stone.

I don't think we need to hold this PR back over some slightly inconsistent, but not actually conflicting, terminology in a comment. If "mock" works well here, there's no reason we can't adopt it for the C side either.

**Comment by sulix** [comment]
> I don't think we need to constraint the C side here. So unless it is a blocker, or something that is anyway an improvement for both sides (like the `const` correctness one I sent once), I would say we should try to find the best way for the C side, and then see if we can work with that.

Agreed: we're still very much experimenting with the C version here. (Whatever we get will just be a collection of function pointers in some sense), so I doubt it'd cause serious issues.

That being said, I'd really like it if, in general, we can keep the C and Rust versions of the KUnit API as similar as possible. And since Rust has a lot of constraints to make APIs safe, we're more than willing to adjust the C API to make it more "Rust friendly" where it makes sense to do so.

File: rust/kernel/kunit.rs (hunk 3 above)
Comment (sulix) [reply rust/kernel/kunit.rs]: Cheers! This patchset now looks good to me.

**Comment by sulix** [comment]
This patchset is (with the possible exception of patch 1, which I probably shouldn't review as the co-author :-P)
```
Reviewed-by: David Gow <davidgow@google.com>
```

Assuming the Rust experts don't find anything wrong with it, I'm happy for it to go in as-is.

Thanks!
— David

File: rust/kernel/kunit.rs [hunk 6]
```diff
@@ -88,3 +88,89 @@ macro_rules! kunit_assert_eq {
...
+///
+/// static mut KUNIT_TEST_CASE: crate::bindings::kunit_case = crate::kunit_case!(name, test_fn);
+/// static mut KUNIT_NULL_CASE: crate::bindings::kunit_case = crate::kunit_case!();
+/// static mut KUNIT_TEST_CASES: &mut[crate::bindings::kunit_case] = unsafe {
+///     &mut[KUNIT_TEST_CASE, KUNIT_NULL_CASE]
+/// };
+/// crate::kunit_test_suite!(suite_name, KUNIT_TEST_CASES);
+/// ```
+#[macro_export]
+macro_rules! kunit_test_suite {
```
Comment (bjorn3) [inline rust/kernel/kunit.rs]: This macro is not sound. It allows passing in a test which does `*std::ptr::null_mut() = 42;` which is UB. It may be possible to use some wrapper types with invariants to make this safe. If not, please add `unsafe` to the macro name.

File: rust/kernel/kunit.rs [hunk 7]
```diff
@@ -88,3 +88,89 @@ macro_rules! kunit_assert_eq {
...
+    ($name:ident, $test_cases:ident) => {
+        static KUNIT_TEST_SUITE_NAME: [i8; 256] = {
```
Comment (bjorn3) [inline rust/kernel/kunit.rs]: Maybe wrap the macro body in `const _: () = { ... };` to prevent these statics from being accessed outside of this macro. Should also clean up the autocompletion list a bit.

**Comment by ojeda** [comment]
> That being said, I'd really like it if, in general, we can keep the C and Rust versions of the KUnit API as similar as possible. 

Definitely! I was just trying to give you as much freedom as possible :)

File: rust/kernel/kunit.rs (hunk 6 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs]: Since this macro is meant to be used mainly by the `kunit_tests` macro, I simply renamed it to `kunit_unsafe_test_suite`. I don't think that adding extra complexity in a macro that is not going to have many users is worth it.

File: rust/kernel/kunit.rs (hunk 7 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs]: Done, thanks for the tip.

**Comment by JoseExposito** [comment]
Hi again, and thanks a lot for all of your comments.

I pushed v3 addressing them:

- Add Reviewed-by: David Gow <davidgow@google.com>
- Rename `kunit_test_suite` to `kunit_unsafe_test_suite` and document the safety requirements
- Wrap `kunit_unsafe_test_suite` body's in `const _: () = { ... };`
- Make the kernel crate available to `kunit_tests`

**Comment by sulix** [comment]
v3 still works for me, thanks!

I'd've had a slight preference for maintaining the old ``kunit_test_suite`` name, instead of renaming it to ``kunit_unsafe_test_suite``, so that it matches the C API, and just marking the macro unsafe, but it looks like that's not possible. So I'm okay with it as-is.

Thanks again for working on this!

**Comment by JoseExposito** [comment]
Thanks a lot David!

After rebasing a branch on this one I noticed that the build fails when `CONFIG_KUNIT` was not set, so I removed the warning in `rust/kernel/lib.rs`:

```diff
+#[allow(unused_extern_crates)]
 extern crate self as kernel;
```

So I fixed it in v4. Sorry for the noise.

File: rust/kernel/kunit.rs [hunk 8]
```diff
@@ -88,3 +88,95 @@ macro_rules! kunit_assert_eq {
...
+                    debugfs: core::ptr::null_mut(),
+                    log: core::ptr::null_mut(),
+                    suite_init_err: 0,
+                };
+
+            // SAFETY: `KUNIT_TEST_SUITE` is static.
+            #[used]
+            #[link_section = ".kunit_test_suites"]
+            static mut KUNIT_TEST_SUITE_ENTRY: *const $crate::bindings::kunit_suite =
+                unsafe { &KUNIT_TEST_SUITE };
```
Comment (bjorn3) [inline rust/kernel/kunit.rs]: If KUnit modifies `KUNIT_TEST_SUITE`, this will need to use `&mut` and then cast to `*mut` first before casting to `*const` to avoid UB. If it doesn't get modified it could use `static` instead of `static mut`.

File: rust/kernel/kunit.rs [hunk 9]
```diff
@@ -88,3 +88,95 @@ macro_rules! kunit_assert_eq {
...
+/// static mut KUNIT_TEST_CASES: &mut[crate::bindings::kunit_case] = unsafe {
+///     &mut[KUNIT_TEST_CASE, KUNIT_NULL_CASE]
+/// };
+/// crate::kunit_unsafe_test_suite!(suite_name, KUNIT_TEST_CASES);
+/// ```
+#[macro_export]
+macro_rules! kunit_unsafe_test_suite {
+    ($name:ident, $test_cases:ident) => {
+        const _: () = {
+            static KUNIT_TEST_SUITE_NAME: [i8; 256] = {
```
Comment (bjorn3) [inline rust/kernel/kunit.rs:151]: This can be `const`, but keeping it as `static` is fine foo.

File: rust/kernel/kunit.rs [hunk 10]
```diff
@@ -183,11 +185,86 @@ macro_rules! kunit_unsafe_test_suite {
...
+/// fn get_boot_ns() -> u64 {
+///     unsafe { bindings::ktime_get_boot_fast_ns() }
+/// }
+///
+/// let time = get_boot_ns();
+/// assert_eq!(time, 1234);
+/// ```
+pub fn in_kunit_test() -> bool {
+    if cfg!(CONFIG_KUNIT) {
+        let test = unsafe { (*Task::current().deref().0.get()).kunit_test };
```
Comment (bjorn3) [inline rust/kernel/kunit.rs:249]: Needs a safety comment.

File: rust/kernel/kunit.rs (hunk 8 above)
Comment (sulix) [reply rust/kernel/kunit.rs]: KUnit does modify instances of ``struct kunit_suite``, but only "private" members (which, being in C, is [only enforced by a comment](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/kunit/test.h#n174)).

So, at least in theory, no rust code should see or care about any changes here (indeed, Rust code really shouldn't be doing anything with this struct other than defining it in the correct linker section, then never reading or writing it again), but technically the struct is mutated.

I've not yet been able to wrap my head completely around Rust rules for partial and/or interior mutability (and I imagine that's more complicated when mixed with auto-generated bindings), but I suspect that's the path a perfect solution would lie down. Given this is used in such limited cases, though, it's probably not worth doing anything overcomplicated.

**Comment by sulix** [comment]
v4 still works for me:
- ``./tools/testing/kunit/kunit.py run --arch x86_64 --kconfig_add CONFIG_RUST=y --make_options LLVM=1 rust_kernel_kunit`` passes both tests.
- Both tests also pass on UML with the appropriate patches backported from the UML tree.
- ``make defconfig`` + ``CONFIG_RUST=y`` (- the DRM subsystem, so it'd build quicker) still builds.
- ``make tinyconfig`` (no Rust support) still builds.

Thanks again for continuing to push this!

File: rust/kernel/kunit.rs (hunk 8 above)
Comment (bjorn3) [reply rust/kernel/kunit.rs]: If you take an immutable reference you are promising rustc that the referenced memory won't be modified until the reference ends. As such rustc would technically be allowed to demote the `static mut` to `static` (it doesn't currently to be clear, but it may in the future). By first taking a mutable reference and then casting it to a *mut raw pointer and then *const raw pointer you avoid this problem. Not that directly casting from &mut to *const will create an intermediate immutable reference somewhat confusingly. Another option is to use `addr_of_mut!(KUNIT_TEST_SUITE) as *const _` to avoid all references.

Comment (ojeda) [reply rust/kernel/kunit.rs]: Yeah, please use `addr_of*!` -- it is what we want to express here, anyway: a pointer, rather than a borrow.

Actually, is there a reason why `addr_of!` would not work? Then we can avoid the `*const` cast too.

Comment (ojeda) [reply rust/kernel/kunit.rs]: By the way:

```rust
static _A: i32 = 0;
static mut _B: *mut i32 = &_A as *const _ as *mut _;
static mut _C: *mut i32 = core::ptr::addr_of!(_A) as *mut _;
static mut _D: *mut i32 = core::ptr::addr_of_mut!(_A); // Error.
```

Is this expected to work @bjorn3? Given the current docs of `addr_of_mut!` (i.e. ignoring the implementation), I would expect so.

However, due to how the macro is implemented via `&raw`, it does not, and then the details leak into the compiler output even in the stable compiler, including mentioning raw references which would be unknown for stable users, as well as claiming a mutable borrow is attempted (but the user was using `addr_of*!` precisely to avoid borrows).

Comment (bjorn3) [reply rust/kernel/kunit.rs]: `addr_of!` produces a raw pointer through which you can't write, while `addr_of_mut!` produces one through which you can write. As such `addr_of_mut!` is necessary here. It indeed seems to not be allowed in statics though. With `-Zteach` it gives an extra note:

```
error[E0764]: raw mutable references are not allowed in the final value of statics
 --> <anon>:4:38
  |
4 | static mut BAR: *const u8 = unsafe { std::ptr::addr_of_mut!(FOO) as *const u8 };
  |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: References in statics and constants may only refer to immutable values.\n\n
          Statics are shared everywhere, and if they refer to mutable data one might violate memory
          safety since holding multiple mutable references to shared data is not allowed.\n\n
          If you really want global mutable state, try using static mut or a global UnsafeCell.
  = note: this error originates in the macro `std::ptr::addr_of_mut` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0764`.
```

In other words using `UnsafeCell` in `KUNIT_TEST_SUITE` is probably the best way to handle this.

Comment (ojeda) [reply rust/kernel/kunit.rs]: `addr_of!`'s docs do not say that the produced pointer cannot be used for writing, though. I was aware of the issue of taking borrows to create pointers, but when using `addr_of!` directly to produce a `*const`, it would be fairly surprising that one is introducing extra promises. What am I missing? (i.e. how can one reach that conclusion? Or are you only talking about the current implementation of the macro?).

Comment (bjorn3) [reply rust/kernel/kunit.rs]: From the bottom of the documentation of `addr_of!`:

> See [addr_of_mut](https://doc.rust-lang.org/stable/std/ptr/macro.addr_of_mut.html) for how to create a pointer to unininitialized data. Doing that with addr_of would not make much sense since **one could only read the data**, and that would be Undefined Behavior.

Comment (ojeda) [reply rust/kernel/kunit.rs]: I saw that, but that is just an example. Not only that, my reading of that is "this is a `*const` pointer, thus the type literally does not allow writing via e.g. `*p = 1;`", not that nobody else couldn't write into that place elsewhere, or that one couldn't cast it into a mutable one, initialize, and then start reading from the original pointer.

For instance, https://doc.rust-lang.org/stable/core/mem/union.MaybeUninit.html#method.as_ptr explicitly mentions writing through the returned pointer is UB (i.e. even if casted etc.), but `addr_of!` does not mention that.

Comment (bjorn3) [reply rust/kernel/kunit.rs]: I found the following places assuming it is UB:

* `addr_of!()` is considered non-mutating unlike `addr_of_mut!()`: https://github.com/rust-lang/rust/blob/2d91939bb7130a8e6c092a290b7d37f654e3c23c/compiler/rustc_middle/src/mir/visit.rs#L665
* `addr_of!()` takes a non-mutable place: https://github.com/rust-lang/rust/blob/2d91939bb7130a8e6c092a290b7d37f654e3c23c/compiler/rustc_mir_build/src/build/expr/into.rs#L304

On the other hand https://github.com/rust-lang/rust/pull/90413 made other code to assume that it may be mutated because as it turns out it hasn't yet been decided if mutating through `addr_of!()` is allowed or not: https://github.com/rust-lang/rust/issues/56604

Comment (ojeda) [reply rust/kernel/kunit.rs]: That is great, thanks a ton for those.

Comment (ojeda) [reply rust/kernel/kunit.rs]: I was reading through those, and saw that https://github.com/rust-lang/unsafe-code-guidelines/issues/257 also discusses the `addr_of!` question (and you replied there! :)

Comment (bjorn3) [reply rust/kernel/kunit.rs]: > (and you replied there! :)

It's hard to keep track of every discussion I ever replied on :)

File: rust/kernel/kunit.rs (hunk 10 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs:249]: Added, thanks!

File: rust/kernel/kunit.rs (hunk 8 above)
Comment (JoseExposito) [reply rust/kernel/kunit.rs]: Changed to `static mut KUNIT_TEST_SUITE: core::cell::UnsafeCell<$crate::bindings::kunit_suite>`

**Comment by JoseExposito** [comment]
Hi everyone,

Sorry for the long delay addressing the 2 remaining issues. I've being swamped with work since the year started.

I pushed v5 adding a missing unsafe comment and declaring `KUNIT_TEST_SUITE` as `UnsafeCell` to avoid UB.

Thanks a lot to everyone for your patience and feedback.

**Comment by sulix** [comment]
FYI, now that the doctests support has landed in the [kselftest/kunit branch](https://git.kernel.org/pub/scm/linux/kernel/git/shuah/linux-kselftest.git/log/?h=kunit), I've gone ahead and rebased this and sent it out:
https://lore.kernel.org/linux-kselftest/20230720-rustbind-v1-0-c80db349e3b5@google.com/T/

With some luck, we'll be able to get it in 6.6.

**Comment by JoseExposito** [comment]
Thanks for rebasing and sending upstream the patches David :D

File: rust/kernel/kunit.rs (hunk 8 above)
Comment (RalfJung) [reply rust/kernel/kunit.rs]: Since `KUNIT_TEST_SUITE_ENTRY` is a pointer you intent to write through, using `*mut` instead of `*const` would make that intent more clear.

(Just stumbled across this while following backrefs from the various issues you linked above. ;)

Also FWIW using both `static mut` and `UnsafeCell` is unnecessary -- but in this case, using `static` + `UnsafeCell` would complain about the lack of `Sync`, so you'd need a `SyncUnsafeCell` type, and using `static mut` directly you need the unstable feature `const_mut_refs` to compute the initial value for `KUNIT_TEST_SUITE_ENTRY`. So  `static mut` + `UnsafeCell`  is a useful hack right now to work around this, but once const_mut_refs is stable you should be able to remove the `UnsafeCell` and instead use `addr_of_mut!` to directly construct a `*mut`.

**Comment by JoseExposito** [comment]
Closing this MR now that, thanks to the big effort from @sulix, this is available upstream.

Thanks a lot for the huge effort upstreaming this David!
